samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+ *
+ * TODO: Implement expiration of entries and triggers
+ */
+public class WindowOperatorImpl<M extends MessageEnvelope, K, WK, WV, WM extends WindowPane<K, WV>> extends OperatorImpl<M, WM> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(WindowOperatorImpl.class);
 
-  public WindowOperatorImpl(WindowOperatorSpec spec, MessageStreamImpl<M> source, Config config, TaskContext context) {
-    // source, config, and context are used to initialize the window kv-store
+  private final PriorityQueue<TriggerTimerState> pendingCallbacks = new PriorityQueue<>();
+  private final WindowInternal<M, K, WV> window;
+  private final KeyValueStore<WindowKey<K>, WV> store = new InternalInMemoryStore<>();
+
+  private final Map<TriggerKey<K>, TriggerImplWrapper> triggers = new HashMap<>();
+
+  private enum TriggerType { EARLY, DEFAULT, LATE }
+
+  public WindowOperatorImpl(WindowOperatorSpec<M, WK, WV> spec) {
     window = spec.getWindow();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    WindowKey<K> storeKey =  getStoreKey(message);
+    BiFunction<M, WV, WV> foldFunction = window.getFoldFunction();
+    WV wv = store.get(storeKey);
+
+    if (wv == null) {
+      wv = window.getInitializer().get();
+    }
+
+    WV newVal = foldFunction.apply(message, wv);
+    store.put(storeKey, newVal);
+
+    if (window.getEarlyTrigger() != null) {
+      TriggerKey<K> triggerKey = new TriggerKey<>(TriggerType.EARLY, storeKey);
+      getOrCreateTriggerWrapper(triggerKey, window.getEarlyTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+
+    if (window.getDefaultTrigger() != null) {
+      TriggerKey<K> triggerKey = new TriggerKey<>(TriggerType.DEFAULT, storeKey);
+      getOrCreateTriggerWrapper(triggerKey, window.getDefaultTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long now = System.currentTimeMillis();
+    TriggerTimerState state;
+    while ((state = pendingCallbacks.peek()) != null && state.getScheduleTimeMs() < now) {
+      pendingCallbacks.remove();
+      state.getCallback().run();
+
+      TriggerImplWrapper wrapper = triggers.get(state.triggerKey);
+      if (wrapper != null) {
+        wrapper.onTimer(collector, coordinator);
+      }
+    }
+  }
+
+  private WindowKey<K> getStoreKey(M message) {
+    Function<M, K> keyExtractor = window.getKeyExtractor();
+    K key = null;
+
+    if (keyExtractor != null) {
+      key = keyExtractor.apply(message);
+    }
+
+    String windowId = null;
+    if (window.getWindowType() == WindowType.TUMBLING) {
+      long triggerDurationMs = ((TimeTrigger<M>) window.getDefaultTrigger()).getDuration().toMillis();
+      final long now = System.currentTimeMillis();
+      Long windowBoundary = now - now % triggerDurationMs;
+      windowId = windowBoundary.toString();
+    }
```

Comment:
For a processing time session window, there will be atmost one entry in the store for every key `K`. However, when we emit the output, we will lift the `K` and add a timestamp of the first message as the `paneId`. 


>> How about count-based window?
I don't think view "count-based window" as a top-level concept. We can still have a tumbling time window with a `count-based` trigger. The current implementation allows for that.

>> I think that at least it makes sense to move this to a getWindowId(M message) function, which can be made abstract later 

For now, there are only two window types (and we have generalized windows abstractly using triggers as much as we can) and a simple `if` is probably sufficient. We can (and probably will) revisit this when we have more window types and event time. It will also be worth to think "Should different window types emit outputs differently?".